### PR TITLE
Turtle functionality in RemoteControl

### PR DESCRIPTION
**Pull request by gabryelreyes**
Added SerialMuxProt channel for turtle application. Makes calls to DifferentialDrive to update the linear and angular speeds.

This PR is a draft as it has not been completely tested

### CODE REVIEW

File: lib/MainNative/src/main.cpp [hunk 1]
```diff
@@ -263,6 +265,10 @@ static int handleCommandLineArguments(PrgArguments& prgArguments, int argc, char
             {
                 prgArguments.serialTxChannel = optarg;
             }
+            else if (0 == strcmp(LONG_OPTIONS[optionIndex].name, "cwd"))
+            {
+                chdir(optarg);
```
Comment (BlueAndi) [inline lib/MainNative/src/main.cpp:270]: Handle it like for others too. Actions derived from it shall be done outside of this function.

File: lib/MainNative/src/main.cpp [hunk 2]
```diff
@@ -307,6 +313,7 @@ static int handleCommandLineArguments(PrgArguments& prgArguments, int argc, char
         printf("\t--serialTxCh <CHANNEL>\t\tSet serial tx channel (ZumoComSystem)."); /* Serial txchannel */
         printf(" Default: %s\n", PRG_ARG_SERIAL_TX_CH_DEFAULT); /* Serial tx channel default value */
         printf("\t-v\t\t\tVerbose mode. Default: Disabled\n");  /* Flag */
+        printf("\t--cwd <dir>\t\tSpecify working directory.");  /* Set process working directory */
```
Comment (BlueAndi) [inline lib/MainNative/src/main.cpp:316]: showPrgArguments() is not updated correspondingly.

File: lib/APPRemoteControl/src/App.cpp [hunk 3]
```diff
@@ -384,3 +386,33 @@ void App_statusChannelCallback(const uint8_t* payload, const uint8_t payloadSize
...
+        int16_t centerSpeed = Util::millimetersPerSecondToStepsPerSecond(turtleSpeedData->linearCenter);
+
+        /* Linear speed is set first. Overwrites all speed setpoints. */
+        diffDrive.setLinearSpeed(centerSpeed);
+
+        /* Angular speed is set on-top of the linear speed. Must be called after setLinearSpeed(). */
+        diffDrive.setAngularSpeed(angularSpeed);
+
+        /* Turtle expects no initial data. Can be called without side-effects when no longer in StartupState. */
+        StartupState::getInstance().notifyInitialDataIsSet();
```
Comment (BlueAndi) [inline lib/APPRemoteControl/src/App.cpp:416]: Why is this necessary? BTW in the driving state the first set speed will be overwritten!

File: lib/APPRemoteControl/src/SerialMuxChannels.h [hunk 4]
```diff
@@ -205,6 +211,13 @@ typedef struct _LineSensorData
     uint16_t lineSensorData[5U]; /**< Line sensor data [digits] normalized to max 1000 digits. */
 } __attribute__((packed)) LineSensorData;
 
+/** Struct of the "Turtle" channel payload. */
+typedef struct _TurtleSpeed
+{
+    int32_t linearCenter; /**< Linear speed of the vehicle center. [mm/s] */
+    int32_t angular;      /**< Angular speed. [mrad/s] */
+} __attribute__((packed)) TurtleSpeed;
```
Comment (BlueAndi) [inline lib/APPRemoteControl/src/SerialMuxChannels.h:219]: Couldn't that be combined with  SpeedData?
SpeedData comes in with mm/s according to structure definition, but its not converted to steps/s!
And SpeedData parameter "center" is not used.

File: lib/APPRemoteControl/src/App.cpp [hunk 5]
```diff
@@ -384,3 +386,33 @@ void App_statusChannelCallback(const uint8_t* payload, const uint8_t payloadSize
...
+    {
+        const TurtleSpeed* turtleSpeedData = reinterpret_cast<const TurtleSpeed*>(payload);
+        DifferentialDrive& diffDrive       = DifferentialDrive::getInstance();
+        int16_t            angularSpeed    = static_cast<int16_t>(turtleSpeedData->angular);
+
+        /* Convert to [steps/s] */
+        int16_t centerSpeed = Util::millimetersPerSecondToStepsPerSecond(turtleSpeedData->linearCenter);
+
+        /* Linear speed is set first. Overwrites all speed setpoints. */
+        diffDrive.setLinearSpeed(centerSpeed);
```
Comment (BlueAndi) [inline lib/APPRemoteControl/src/App.cpp:410]: Interfaces are mixed now ... here diffDrive is used and for SpeedData the target speed is set via DrivingState::setTargetSpeeds().

**Review by BlueAndi — CHANGES_REQUESTED** [review]

File: lib/APPRemoteControl/src/SerialMuxChannels.h (hunk 4 above)
Comment (gabryelreyes) [reply lib/APPRemoteControl/src/SerialMuxChannels.h:219]: It could be combined. I decided against it in order to maintain SpeedData untouched for the other applications that use it. Should I then add the angular speed to SpeedData?

File: lib/APPRemoteControl/src/App.cpp (hunk 3 above)
Comment (gabryelreyes) [reply lib/APPRemoteControl/src/App.cpp:416]: It is necessary to get into DrivingState, as the StartUpState waits for initial data such as Position coordinates. This comes from the ConvoyFollower Application in DCS, but its not necessary for the TurtleSim

File: lib/APPRemoteControl/src/SerialMuxChannels.h (hunk 4 above)
Comment (BlueAndi) [reply lib/APPRemoteControl/src/SerialMuxChannels.h:219]: Hmm ... shall we provide two channels?
1. linear speed left and right (no center, no angular)
2. linear speed center + angular speed

And both in mm/s?

File: lib/APPRemoteControl/src/App.cpp (hunk 3 above)
Comment (BlueAndi) [reply lib/APPRemoteControl/src/App.cpp:416]: But the ConvoyFollower is a different app, therefore its not necessary, or?

File: lib/APPRemoteControl/src/SerialMuxChannels.h (hunk 4 above)
Comment (gabryelreyes) [reply lib/APPRemoteControl/src/SerialMuxChannels.h:219]: The second channel is this "TurtleSpeed" channel. At least that was the idea

Comment (BlueAndi) [reply lib/APPRemoteControl/src/SerialMuxChannels.h:219]: I would keep RemoteControl as generic as possible. This way other applications can use the interfaces without the necessity to define their own channel. That means we have two channels, one is dealing with motor speeds (SpeedData) and the other with robot speeds (TurtleSpeed).

It makes sense to rename the channels and the callbacks, to show this to the user.
E.g. MotorSpeed / RobotSpeed channels.

File: lib/APPRemoteControl/src/App.cpp (hunk 3 above)
Comment (gabryelreyes) [reply lib/APPRemoteControl/src/App.cpp:416]: Oh. You are right, sorry. Will change this

File: lib/APPRemoteControl/src/App.cpp [hunk 6]
```diff
@@ -363,9 +361,9 @@ static void App_cmdChannelCallback(const uint8_t* payload, const uint8_t payload
 void App_motorSpeedSetpointsChannelCallback(const uint8_t* payload, const uint8_t payloadSize, void* userData)
 {
     (void)userData;
-    if ((nullptr != payload) && (SPEED_SETPOINT_CHANNEL_DLC == payloadSize))
+    if ((nullptr != payload) && (MOTOR_SPEED_SETPOINT_CHANNEL_DLC == payloadSize))
     {
-        const SpeedData* motorSpeedData = reinterpret_cast<const SpeedData*>(payload);
+        const MotorSpeed* motorSpeedData = reinterpret_cast<const MotorSpeed*>(payload);
```
Comment (BlueAndi) [inline lib/APPRemoteControl/src/App.cpp:366]: Conversion to steps/s is missing.

File: lib/APPRemoteControl/src/App.cpp [hunk 7]
```diff
@@ -390,16 +388,16 @@ void App_statusChannelCallback(const uint8_t* payload, const uint8_t payloadSize
 /**
  * Receives Turtle speed setpoints over SerialMuxProt channel.
  *
- * @param[in] payload       Linear and angular speed setpoints in a TurtleSpeed structure.
- * @param[in] payloadSize   Size of the TurtleSpeed structure.
+ * @param[in] payload       Linear and angular speed setpoints in a RobotSpeed structure.
+ * @param[in] payloadSize   Size of the RobotSpeed structure.
  * @param[in] userData      Instance of App class.
  */
 void App_turtleChannelCallback(const uint8_t* payload, const uint8_t payloadSize, void* userData)
```
Comment (BlueAndi) [inline lib/APPRemoteControl/src/App.cpp]: Change callback name.

**Review by BlueAndi — CHANGES_REQUESTED** [review]

File: lib/APPRemoteControl/src/App.cpp (hunk 7 above)
Comment (gabryelreyes) [reply lib/APPRemoteControl/src/App.cpp]: Your comment is not on the latest commit, where this change was already fixed

**Review by BlueAndi — APPROVED** [review]
